stdeb/pypi_simple.py

```diff
@@ -0,0 +1,46 @@
+import re
+import requests
+import stdeb
+
+USER_AGENT = "pypi-install/%s ( https://github.com/astraw/stdeb )" % \
+    stdeb.__version__
+
+
+def normalize_package_name(package_name):
+    return re.sub(r"[-_.]+", "-", package_name).lower()
+
+
+class PyPIClient(object):
+    def __init__(self, pypi_url="https://pypi.org", user_agent=USER_AGENT):
+        self.pypi_url = pypi_url
+        self._http_session = requests.Session()
+        self._http_session.headers["User-Agent"] = user_agent
+
+    def release_versions(self, package_name, current_only=False):
+        package_name = normalize_package_name(package_name)
+        if current_only:
+            response = self._http_session.get("%s/pypi/%s/json" % (self.pypi_url, package_name))
+            data = response.json()
+            return data['info']['version']
+        else:
+            response = self._http_session.get(
+                "%s/simple/%s" % (self.pypi_url, package_name),
+                headers={"Accept": "application/vnd.pypi.simple.latest+json"})
+            data = response.json()
+            return data["versions"]
+
+    def download_url(self, package_name, version):
+        download_url = None
+        md5_digest = None
+        package_name = normalize_package_name(package_name)
+        response = self._http_session.get(
+            "%s/pypi/%s/%s/json" % (self.pypi_url, package_name, version))
+        data = response.json()
+        for url in data["urls"]:
+            if url["packagetype"] == "sdist" and \
+                    url["python_version"] == "source" and \
+                    url["url"].endswith((".tar.gz", ".zip")):
+                download_url = url["url"]
+                md5_digest = url["digests"]["md5"]
```

Comment:
The original algorithm has a `break` here, meaning that the first encountered match is used. This continues to loop, meaning that the last match is used. Is that an intentional deviation?

Reply:
Good catch! No it isn't. I don't know that there would be a situation with multiple sdist payloads for a given version so I think the `break` _should_ just be a minor optimization, but it is worth preserving the behavior. In the event that there are multiple sdist urls I'm not sure that the sorting from the server is reliable so we just have to trust that the first one we get is good enough.

Reply:
Restored the `break` in 86d8edc